cabal-install/src/Distribution/Client/Config.hs

```diff
@@ -587,40 +590,62 @@ initialSavedConfig = do
     }
   }
 
-defaultCabalDir :: IO FilePath
-defaultCabalDir = getAppUserDataDirectory "cabal"
-
-getCabalDir :: IO FilePath
-getCabalDir = do
-  mDir <- lookupEnv "CABAL_DIR"
+-- The default behaviour of cabal-install is to use the XDG directory
+-- standard.  However, if CABAL_DIR is set, we instead use that
+-- directory as a single store for everything cabal-related, like the
+-- old ~/.cabal behaviour.  Also, for backwards compatibility, if
+-- ~/.cabal exists we treat that as equivalent to CABAL_DIR being set.
+-- This function abstracts that decision-making.
+getDefaultDir :: XdgDirectory -> FilePath -> IO FilePath
+getDefaultDir xdg subdir = do
+  mDir <- maybeGetCabalDir
   case mDir of
-    Nothing -> defaultCabalDir
-    Just dir -> return dir
+    Just dir -> return $ dir </> subdir
+    Nothing -> getXdgDirectory xdg $ "cabal" </> subdir
 
 defaultConfigFile :: IO FilePath
-defaultConfigFile = do
-  dir <- getCabalDir
-  return $ dir </> "config"
+defaultConfigFile =
+  getDefaultDir XdgConfig "config"
 
 defaultCacheDir :: IO FilePath
-defaultCacheDir = do
-  dir <- getCabalDir
-  return $ dir </> "packages"
+defaultCacheDir =
+  getDefaultDir XdgCache "packages"
```

Comment:
E.g. this one shouldn't look at whether `~/.cabal` folder exists. The default cache dir should be just `getXdgDirectory </> "cabal" </> "packages"`.

Reply:
... but as I tried to say previously, I'm not sure if these functions are **also** used to fill missed fields in cabal `config`. I think that is bad functionality, and `cabal` should rather just fail (and separately provide a way to update the config, there is `cabal user-config update` after all).

Reply:
These functions are used to fill the missing fields in the config file.  It may be bad functionality, but I tried to preserve complete backwards compatibility in this PR.  In particular, if ~/.cabal already exists or `CABAL_DIR` is set, there will be no change in behaviour.  Changing the interpretation of missing config fields is a different feature, and can be done either before or after this PR, but it would not affect the ease of implementing XDG compliance.